File: Depth-first_search(DFS)/python/04_MooyoMooyo.py
# fast campus 강의
# https://www.acmicpc.net/problem/16768
# 2

N, K = map(int, input().split())
matrix = [list(input()) for _ in range(N)]
visited = [[False] * 10 for _ in range(N)]
visited2 = [[False] * 10 for _ in range(N)]
mx = [-1, 1, 0, 0]
my = [0, 0, -1, 1]

def dfs(x, y, cnt):
    visited[x][y] = True
    for i in range(4):
        nx = x + mx[i]
        ny = y + my[i]
        if nx < 0 or nx >= N or ny < 0 or ny >= 10:
            continue
        if not(visited[nx][ny]) and matrix[x][y] == matrix[nx][ny]:
            cnt = dfs(nx, ny, cnt + 1)
    return cnt

def dfs2(x, y):
    visited2[x][y] = True
    for i in range(4):
        nx = x + mx[i]
        ny = y + my[i]
        if nx < 0 or nx >= N or ny < 0 or ny >= 10:
            continue
        if not(visited2[nx][ny]) and matrix[x][y] == matrix[nx][ny]:
            dfs2(nx, ny)
    matrix[x][y] = '0'

def down():
    for y in range(10):
        tmp_list = []
        for x in range(N):
            if matrix[x][y] != '0':
                tmp_list.append(matrix[x][y])
        for x in range(N - len(tmp_list)):
            matrix[x][y] = '0'
        for x in range(len(tmp_list)):
            matrix[x + N - len(tmp_list)][y] = tmp_list[x]

while 1:
    chk = False
    visited = [[False] * 10 for _ in range(N)]
    visited2 = [[False] * 10 for _ in range(N)]
    for i in range(N):
        for j in range(10):
            if not(visited[i][j]) and matrix[i][j] != '0':
                cnt = dfs(i, j, 1)
                if cnt >= K:
                    chk = True
                    dfs2(i, j)
    if not(chk):
        break
    down()

for i in matrix:
    print(''.join(i))

File: Depth-first_search(DFS)/c++/04_MooyoMooyo.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/16768
// 1

#include <iostream>
#include <queue>
#include <cstring>

using namespace std;

int SIZE = 10;
int matrix[105][12];
bool visited[105][12], visited2[105][12];
int mx[4] = { -1, 1, 0, 0 };
int my[4] = { 0, 0, -1, 1 };
int N, K;

int dfs(int x, int y, int count)
{
	visited[x][y] = true;
	for (int i = 0; i < 4; i++) {
		int nx = x + mx[i];
		int ny = y + my[i];

		if (nx <= 0 || nx > N || ny <= 0 || ny > SIZE)
			continue;
		if (!visited[nx][ny] && matrix[nx][ny] == matrix[x][y])
			count = dfs(nx, ny, count + 1);
	}
	return count;
}

void dfs(int x, int y)
{
	visited2[x][y] = true;
	for (int i = 0; i < 4; i++) {
		int nx = x + mx[i];
		int ny = y + my[i];

		if (nx <= 0 || nx > N || ny <= 0 || ny > SIZE)
			continue;
		if (!visited2[nx][ny] && matrix[nx][ny] == matrix[x][y])
			dfs(nx, ny);
	}
	matrix[x][y] = 0;
}

void down() {
	for (int y = 1; y <= SIZE; y++) {
		vector<int> tmp;
		for (int x = 1; x <= N; x++)
			if (matrix[x][y])
				tmp.push_back(matrix[x][y]);
		for (int x = 1; x <= N - tmp.size(); x++)
			matrix[x][y] = 0;
		for (int x = 0; x < tmp.size(); x++)
			matrix[N - tmp.size() + x + 1][y] = tmp[x];
	}
}

int main()
{
	cin >> N >> K;
	for (int i = 1; i <= N; i++)
		for (int j = 1; j <= SIZE; j++)
			scanf("%1d", &matrix[i][j]);

	while (1) {
		bool chk = true;
		memset(visited, false, sizeof(visited));
		memset(visited2, false, sizeof(visited2));
		for (int i = 1; i <= N; i++) {
			for (int j = 1; j <= SIZE; j++) {
				if (matrix[i][j] != 0 && !visited[i][j]) {
					int cnt = dfs(i, j, 1);
					if (cnt >= K) {
						dfs(i, j);
						chk = false;
					}
				}
			}
		}
		if (chk)
			break;
		down();
	}

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= SIZE; j++)
			printf("%d", matrix[i][j]);
		printf("\n");
	}
}